### PR TITLE
Added function to build stategraphs

### PR DESCRIPTION
**Pull request by ptersilie**
This function takes a `Grammar` as input and then builds a `StateGraph` containing all states and their edges.

The `test_stategraph` uses the following grammar:

```
Z : S
S : Sb
    bAa
A : aSc
    a
    aSb
```

the resulting graph looks like this (with one extra state `Start! : Z. {$}` and an additional rule `Start! : .Z {$}` in State #0):
test_grammar
Note that the state numbers may be different each execution due to the `LR1Items` being stored in a `HashSet` and thus being processed in arbitrary order.

### CODE REVIEW

File: src/lib/pgen.rs [hunk 1]
```diff
@@ -344,3 +360,127 @@ pub fn goto1(grm: &Grammar, firsts: &HashMap<String, HashSet<String>>,
     closure1(grm, firsts, &mut goto);
     goto
 }
+
+// Haven't decided yet if I want to use a struct or not
+// struct State {
+//     id: usize,
```
Comment (ltratt) [inline src/lib/pgen.rs]: What would the ID be?

**Comment by ptersilie** [comment]
Here is what the algorithm does:

1) Create a state 0 and build it's closure.
2) For each Symbol `s` next to the `dot` in state 0, use `goto1` to generate the states that follow after the parser reads `s`
3) a) Give the new state a unique state id and create an edge (state0, `s`) -> unique_id
    b) If the state already exists, just create an edge from the current state to the existing one
4) Repeat from step 2) with all newly added states, until no new states are being added

File: src/lib/pgen.rs (hunk 1 above)
Comment (ptersilie) [inline src/lib/pgen.rs]: I was thinking at one point to store the unique_ids (eg. state numbers) inside state. But now their id is just their position in the `states`.

**Comment by ltratt** [comment]
The basic algorithm looks simple enough, but the print stuff looks a bit messy.

I'd be fine merging in a simple version of the algorithm and then working on an optimisation in a branch. However, I think we should clean up the print stuff somehow. It does suggest that maybe there's a need for a struct that can print out a state?

**Comment by ptersilie** [comment]
I used the print to check that the result is correct before I started writing the tests. And just left it in for you so you can see the actual output that the algorithm produces. It can be removed if you have no use for it.

**Comment by ltratt** [comment]
I'm OK with removing the print stuff.

File: src/lib/pgen.rs (hunk 1 above)
Comment (ltratt) [inline src/lib/pgen.rs]: Sounds like a simpler ID to me and saves memory to boot!

**Comment by ltratt** [comment]
Am I right in thinking that, after squashing the last commit, this could sensibly be merged?

**Comment by ptersilie** [comment]
Possibly. Should I leave the commented state struct in?

**Comment by ltratt** [comment]
I'm not sure the commented state struct adds much, so let's kill it.

File: src/lib/pgen.rs (hunk 1 above)
Comment (ptersilie) [inline src/lib/pgen.rs]: Yes, we need a way to get the state from its state number anyway, so storing the number on the state itself is not really necessary.

**Pull request merged**